src/hera/workflows/_mixins.py

```diff
@@ -632,21 +633,39 @@ def __call__(self, *args, **kwargs) -> Union[Step, Task]:
         # the step/task will miss adding them when building the final arguments
         kwargs["arguments"] = arguments
 
-        try:
-            from hera.workflows.steps import Step
-
-            return Step(*args, template=self, **kwargs)
-        except InvalidType:
-            pass
-
-        try:
-            from hera.workflows.task import Task
+        from hera.workflows.dag import DAG
+        from hera.workflows.script import Script
+        from hera.workflows.steps import Parallel, Step, Steps
+        from hera.workflows.task import Task
+        from hera.workflows.workflow import Workflow
+
+        if _context.pieces:
+            if isinstance(_context.pieces[-1], Workflow):
+                # Notes on callable templates under a Workflow:
+                # * If the user calls a script directly under a Workflow (outside of a Steps/DAG) then we add the script
+                #   template to the workflow and return None.
+                # * Containers, ContainerSets and Data objects (i.e. subclasses of CallableTemplateMixin) are already
```

Comment:
This is a question of whether you want to be able to add containers as a template through a callable (I would say no). The options are:

### 1. WAD: use a function to retrieve the container defined elsewhere (but returning a fresh instantation)
```py
def get_container():
    return Container(
    name="whalesay",
    image="docker/whalesay:latest",
    command=["cowsay"],
    args=["{{inputs.parameters.message}}"],
    inputs=Parameter(name="message"),
)

with Workflow(
    generate_name="arguments-parameters-",
    entrypoint="whalesay",
    arguments=Parameter(name="message", value="hello world"),
) as w:
    c = get_container()  # Container is instantiated in the function call and added correctly, c is now callable in steps/dags
```

### 2. Enhancement: make Containers callable under `Workflow`s
```py
whalesay = Container(
    name="whalesay",
    image="docker/whalesay:latest",
    command=["cowsay"],
    args=["{{inputs.parameters.message}}"],
    inputs=Parameter(name="message"),
)

with Workflow(
    generate_name="arguments-parameters-",
    entrypoint="whalesay",
    arguments=Parameter(name="message", value="hello world"),
) as w:
    whalesay()  # Container is already instantiated outside so this doesn't currently work, should this return something or None?
```

I'm hesitant to add that functionality as we somewhat accidentally added it for script-decorated functions. It encourages bad practice (IMO) of global variables - e.g. what if the `whalesay` container is modified in code elsewhere (changing the image etc) without me realising? We should try to stay close to being a definition language (as opposed to going full OOP) as a yaml alternative. The script-decorator works as a workflow-level callable because it's a fresh function call each time. Making the container callable under Workflows also confuses the callable container pattern under steps/dags - should a user be given a container from the call? e.g.:

```py
with Workflow(
    generate_name="arguments-parameters-",
    entrypoint="whalesay",
    arguments=Parameter(name="message", value="hello world"),
) as w:
    container = whalesay()
    with Steps(name="s"):
        # Do I call `container()` or `whalesay()` here?
        whalesay()
        container()
```